### DISCUSSION TITLE
extending useFetch and using $fetch

### DISCUSSION THREAD

**Discussion by acidjazz** [nuxt/nuxt#18138] [General]
As I work on a plugin for my API - I was hoping to integrate and pass through fetch so I can easily add my headers/types/etc.

I have a function for getting an index/listing of data:

```ts
public async index <Results>(endpoint: string, params: SearchParams): Promise<api.MetApiResults & { data: Results }> {
  try {
    return await $fetch<api.MetApiResults & { data: Results }>(endpoint, this.fetchOptions(params))
  } catch (error) {
    for (const err of error.response.data.errors)
      this.$toast.show({
        type: 'danger',
        message: err.detail ?? err.message ?? '',
        timeout: 0,
      })
  }
}
```

This uses fetch as `$fetch` , allows me to properly display errors, and also lets me pass my custom interfaces/arrays so its properly typed out, the idea is I use it on a page like so:

```ts
const result = $api.index<Examples>('/example', { count: 3 })
```

This works well, my problem is I'm seeing this execute both on the server and the client side, calling the endpoint twice and then reporting hydration issues on display (its randomly generated data)

Is this approach not recommended? should I be trying something like 

```ts
const { data: result } = await useAsyncData('examples', () => $api.index<Examples>('/example', { count: 3 }))

```

**Comment by wobsoriano** [comment]
Yeah you need to use useAsyncData or useFetch or put your fetcher inside onMounted or onBeforeMounted instead

**Comment by acidjazz** [comment]
@wobsoriano do you mean use actual `useFetch` in my class? which do you prefer?

**Reply by wobsoriano** [reply]
Use it in `useAsyncData`

```ts
await useAsyncData('examples', () => $api.index<Examples>('/example', { count: 3 }))
```
